### PR TITLE
Teach mage clean

### PR DESCRIPTION
**Pull request by Zanadar**
Addresses https://github.com/magefile/mage/issues/19 

Starts the refactor of `-init` and `-version` into commands.

### CODE REVIEW

**Comment by natefinch** [comment]
Thanks for this!  Is it ready for review, or are you still working?  Let me know.

**Comment by Zanadar** [comment]
It should be ready for a long look. Thanks!

File: mage/main.go [hunk 1]
```diff
@@ -99,33 +111,45 @@ func ParseAndRun(dir string, stdout, stderr io.Writer, stdin io.Reader, args []s
...
 		return 0
+	case CLEAN:
+		dir := mg.CacheDir()
+		if err := removeContents(dir); err != nil {
+			log.Println("Error:", err)
+			return 1
+		}
+		log.Println(dir, "cleaned")
+		return 0
 	}
```
Comment (natefinch) [inline mage/main.go]: can you add 
```
case None:
    return Invoke(inv)
default:
   panic(fmt.Errorf("Unknown command type: %v", cmd)
}
```

That way we will make sure to catch if we add any new commands.

Also, list should be a command (the -l flag)

Comment (Zanadar) [reply mage/main.go]: Would you mind if I kept the updating of `list` to a command as a separate PR?

**Comment by Zanadar** [comment]
@natefinch Updated per your comments.  I wonder if I might leave the updating of `list` to a command to a separate PR to keep this PR focussed?

**Comment by Zanadar** [comment]
@natefinch Anything I should do here to see this get to master?

File: mage/main.go [hunk 2]
```diff
@@ -401,3 +439,22 @@ func RunCompiled(inv Invocation, exePath string) int {
...
 	return sh.ExitStatus(c.Run())
 }
+
+func removeContents(dir string) error {
+	d, err := os.Open(dir)
```
Comment (natefinch) [inline mage/main.go]: why are you opening the directory instead of ioutil.ReadDir(dir) ?

File: mage/main.go [hunk 3]
```diff
@@ -401,3 +439,22 @@ func RunCompiled(inv Invocation, exePath string) int {
...
+	if err != nil {
+		return err
+	}
+	defer d.Close()
+	names, err := d.Readdirnames(-1)
+	if err != nil {
+		return err
+	}
+	for _, name := range names {
+		err = os.RemoveAll(filepath.Join(dir, name))
```
Comment (natefinch) [inline mage/main.go]: if you use ReadDir, then you could check if something was a file before removing it, that way if someone happens to make a subdirectory in this directory, we won't wipe it out.

**Comment by Zanadar** [comment]
Cool! Made some changes per your feedback. LMK what you think. Thanks for your dutiful attention to this humble PR!

**Comment by natefinch** [comment]
LGTM, I'll merge once the tests pass

**Comment by Zanadar** [comment]
Beautiful!

**Comment by natefinch** [comment]
Thanks for all your work! :)

**Pull request merged**